File: app/src/main/java/com/constraint/vagabond/data/RecAreaRepository.java
package com.constraint.vagabond.data;

import android.content.Context;
import android.os.AsyncTask;

import com.constraint.vagabond.data.entities.RecreationalArea;
import com.constraint.vagabond.data.local.RecAreaDao;
import com.constraint.vagabond.data.local.RecAreaDatabase;

import java.util.List;

public class RecAreaRepository {

  private RecAreaDao recAreaDao;
  private List<RecreationalArea> recreationalAreas;
  private RecreationalArea singleArea;

  public RecAreaRepository(Context context) {
    RecAreaDatabase db = RecAreaDatabase.getInstance(context);
    recAreaDao = db.recAreaDao();
    recreationalAreas = recAreaDao.getAreas();
  }

  public List<RecreationalArea> getAreas() {
    return recreationalAreas;
  }

  public RecreationalArea getSingleArea(String recAreaId) {
    return recAreaDao.getArea(recAreaId);
  }

  public void saveAreas(List<RecreationalArea> recreationalAreas) {
    new asyncInsertTask(recAreaDao).execute(recreationalAreas);
  }

  private static class asyncInsertTask extends AsyncTask<List<RecreationalArea>, Void, Void> {
    private RecAreaDao mRecAreaDao;

    asyncInsertTask(RecAreaDao recAreaDao) {
      mRecAreaDao = recAreaDao;
    }

    @Override
    protected Void doInBackground(List<RecreationalArea>... lists) {
      mRecAreaDao.save(lists[0]);
      return null;
    }
  }
}

File: app/src/main/java/com/sierra/vagabond/data/RecAreaRepository.kt
package com.sierra.vagabond.data

import com.sierra.vagabond.api.AreasApiService
import com.sierra.vagabond.api.SierraApiService
import com.sierra.vagabond.data.entities.*
import com.sierra.vagabond.data.local.RecAreaDao
import com.sierra.vagabond.di.AreasAPI
import com.sierra.vagabond.di.SierraAPI
import com.sierra.vagabond.utils.BY_NAME
import com.sierra.vagabond.utils.CAMPING
import javax.inject.Inject
import javax.inject.Singleton


@Singleton
class RecAreaRepository @Inject constructor(private val recAreaDao: RecAreaDao, @AreasAPI private val service: AreasApiService, @SierraAPI private val sierraAPI: SierraApiService) {

    suspend fun getRecAreasList(query: String): RecreationalAreaList {
        return service.getRecreationalAreaData(query = query, full = true, activity = CAMPING, by = BY_NAME)
    }

    suspend fun getSingleArea(recAreaId: String?): RecreationalArea {
        return recAreaDao.getArea(recAreaId)
    }

    suspend fun createWatch(watch: WatchRequest) {
        sierraAPI.createWatch(watch)
    }

    suspend fun insert(recreationalArea: RecreationalArea) {
        recAreaDao.save(recreationalArea)
    }

    suspend fun insertAll(recreationalAreas: List<RecreationalArea>) {
        recAreaDao.saveAll(recreationalAreas)
    }

    suspend fun clearAll() {
        recAreaDao.deleteAll()
    }

    suspend fun registerToken(tokenRequest: TokenRequest) {
        sierraAPI.registerToken(tokenRequest)
    }
}
